File: langs/python/tests/unit/helpers/test_hub.py
"""In-process IMessageHub-equivalent for unit tests."""

from __future__ import annotations

from typing import TYPE_CHECKING, TypeVar

import reactivex as rx
from reactivex.subject import Subject

from vmx.messages.protocols import Message

if TYPE_CHECKING:
    pass

TMessage = TypeVar("TMessage", bound=Message)


class TestHub:
    """Subject-backed test hub. Subscribers can use Rx operators directly.

    This is a stand-alone helper that does NOT inherit from ``MessageHub``
    (the concrete class lands in Task 4).  It satisfies the structural
    ``MessageHubProto`` interface: ``.messages`` and ``.send()``.
    """

    __test__ = False  # tell pytest this is not a test class

    def __init__(self) -> None:
        self._subject: Subject[Message] = Subject()

    @property
    def messages(self) -> rx.Observable[Message]:
        return self._subject

    def send(self, message: TMessage) -> None:
        self._subject.on_next(message)

    def dispose(self) -> None:
        self._subject.on_completed()
        self._subject.dispose()

File: langs/csharp/tests/VMx.Tests/Helpers/TestHub.cs
using System.Reactive.Subjects;
using VMx.Messages;
using VMx.Services;

namespace VMx.Tests.Helpers;

/// <summary>
/// In-process IMessageHub for tests. Backed by a Subject so subscribers
/// can use Rx operators (Where, ObserveOn, etc.) directly.
/// </summary>
public sealed class TestHub : IMessageHub, IDisposable
{
    private readonly Subject<IMessage> _subject = new();

    public IObservable<IMessage> Messages => _subject;

    public void Send<TMessage>(TMessage message) where TMessage : IMessage
        => _subject.OnNext(message);

    public void Dispose() => _subject.Dispose();
}
